ftdc/ftdc.go

```diff
@@ -356,6 +357,131 @@ func (ftdc *FTDC) constructDatum() datum {
+	mapOrderSet := make(map[string]struct{})
+
+	dataMapOrder := make([]string, 0, len(data))
+	var (
+		fields         []string
+		values         []float32
+		iterationOrder []string
+	)
+	if inputSchema != nil {
+		fields = make([]string, 0, len(inputSchema.fieldOrder))
+		values = make([]float32, 0, len(inputSchema.fieldOrder))
+		iterationOrder = make([]string, 0, len(inputSchema.mapOrder))
+
+		for _, key := range inputSchema.mapOrder {
+			iterationOrder = append(iterationOrder, key)
+		}
+	} else {
+		schemaChanged = true
+		iterationOrder = make([]string, 0, len(data))
+		for key := range data {
+			iterationOrder = append(iterationOrder, key)
+		}
+	}
+
+	for _, key := range iterationOrder {
+		mapOrderSet[key] = struct{}{}
+
+		// Walk over the datum in `mapOrder` to ensure we gather values in the order consistent with
+		// the current schema.
+		stats, exists := data[key]
+		if !exists {
+			schemaChanged = true
+			continue
+		}
+
+		itemFields, itemNumbers, err := flattenStructNew(reflect.ValueOf(stats))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		dataMapOrder = append(dataMapOrder, key)
+		for idx := range itemFields {
+			fields = append(fields, fmt.Sprintf("%v.%v", key, itemFields[idx]))
+		}
+		values = append(values, itemNumbers...)
+	}
+
+	// Check for a schema change by walking all of the keys (`Statser`s) in the data and see if
+	// there is anything new.
+	for dataKey, stats := range data {
+		if _, exists := mapOrderSet[dataKey]; exists {
+			continue
+		}
+
+		// We found a statser that did not exist before. Let's add it to our results.
+		schemaChanged = true
+		itemFields, itemNumbers, err := flattenStructNew(reflect.ValueOf(stats))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		dataMapOrder = append(dataMapOrder, dataKey)
+		for idx := range itemFields {
+			fields = append(fields, fmt.Sprintf("%v.%v", dataKey, itemFields[idx]))
+		}
+		values = append(values, itemNumbers...)
+	}
+
+	if inputSchema != nil && !slices.Equal(inputSchema.fieldOrder, fields) {
+		schemaChanged = true
+	}
+
+	if schemaChanged {
+		return &schema{dataMapOrder, fields}, values, nil
+	}
+
+	return inputSchema, values, nil
+}
+
+func (ftdc *FTDC) writeDatumNew(datum datum) error {
+	toWrite, err := ftdc.getWriter()
+	if err != nil {
+		return err
+	}
+
+	// walk will return the schema it found alongside the flattened data. Errors are terminal. If
+	// the schema is the same, the `newSchema` pointer will match `ftdc.currSchema` and `err` will
+	// be nil.
+	newSchema, flatData, err := walk(datum.Data, ftdc.currSchema)
+	if err != nil {
+		return err
+	}
+
+	// The first iteration when `currSchema` is nil, it's guaranteed that the generations will not
+	// be equal. In the happy path where the schema hasn't changed, the `walk` function is
+	// guaranteed to return the same schema object.
+	if datum.generationID != ftdc.outputGenerationID || ftdc.currSchema != newSchema {
```

Comment:
I think generations will go away entirely. But I'll save that for a later commit in this PR